receiver/kymastatsreceiver/kyma_scraper.go

```diff
@@ -0,0 +1,208 @@
+	namespace string
+	kind      string
+
+	state      string
+	conditions []condition
+}
+
+type condition struct {
+	condType string
+	status   string
+	reason   string
+}
+
+type fieldNotFoundError struct {
+	field string
+}
+
+func (e *fieldNotFoundError) Error() string {
+	return fmt.Sprintf("field not found: %s", e.field)
+}
+
+func newKymaScraper(client dynamic.Interface, settings receiver.Settings, moduleGVRs []schema.GroupVersionResource, mbc metadata.MetricsBuilderConfig) (scraperhelper.Scraper, error) {
+	ks := kymaScraper{
+		client:     client,
+		logger:     settings.Logger,
+		moduleGVRs: moduleGVRs,
+		mb:         metadata.NewMetricsBuilder(mbc, settings),
+	}
+
+	return scraperhelper.NewScraper(metadata.Type.String(), ks.scrape)
+}
+
+func (ks *kymaScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	stats, err := ks.collectModuleStats(ctx)
+	if err != nil {
+		return pmetric.Metrics{}, err
+	}
+
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	for _, s := range stats {
+		ks.mb.RecordKymaModuleStatusStateDataPoint(now, int64(1), s.state)
+		rb := ks.mb.NewResourceBuilder()
+		rb.SetK8sNamespaceName(s.namespace)
+		rb.SetKymaModuleName(s.kind)
+		for _, c := range s.conditions {
+			val := conditionStatusToValue(c.status)
+			ks.mb.RecordKymaModuleStatusConditionsDataPoint(now, val, c.reason, c.status, c.condType)
+		}
+		ks.mb.EmitForResource(metadata.WithResource(rb.Emit()))
+	}
+
+	return ks.mb.Emit(), nil
+}
+
+func (ks *kymaScraper) collectModuleStats(ctx context.Context) ([]moduleStats, error) {
+	var res []moduleStats
+	for _, gvr := range ks.moduleGVRs {
+		moduleList, err := ks.client.Resource(gvr).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			ks.logger.Error("Error fetching module list",
+				zap.Error(err),
+				zap.String("group", gvr.Group),
+				zap.String("version", gvr.Version),
+				zap.String("resource", gvr.Resource))
+			return nil, err
+		}
+
+		for _, module := range moduleList.Items {
+			stats, err := ks.unstructuredToStats(module)
+			if err != nil {
+				ks.logger.Error("Error converting unstructured module to stats",
+					zap.Error(err),
+					zap.String("name", module.GetName()),
+					zap.String("namespace", module.GetNamespace()),
+					zap.String("kind", module.GetKind()),
+				)
+				continue
+			}
+
+			res = append(res, *stats)
+
+			// Take only the first valid module custom resource
+			break
+		}
+	}
+
+	return res, nil
+}
+
+func (ks *kymaScraper) unstructuredToStats(module unstructured.Unstructured) (*moduleStats, error) {
+	status, found, err := unstructured.NestedMap(module.Object, "status")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, &fieldNotFoundError{"status"}
+	}
+
+	state, found, err := unstructured.NestedString(status, "state")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, &fieldNotFoundError{"state"}
+	}
+
+	unstructuredConds, found, err := unstructured.NestedSlice(status, "conditions")
```

Comment:
are conditions really mandatory (included in the kyma module contract)?

Reply:
There is no kyma module contract about the status conditions

Reply:
And with that they are not mandatory